### PR TITLE
Add player rank graphics

### PR DESCRIPTION
**Pull request by Epix37**
Not ready to be merged I guess.

- ~~Images are all placeholders~~ 
- Not confident about the css, though it seems to work
- Not sure about using `CardArt` in `RankArt`, though that also seems to work fine.

### CODE REVIEW

**Comment by beheh** [comment]
The `CardArt`/`RankArt`/... stuff isn't really nice anyway, with a lot of redundant prop parsing. At some point I want to change to real inheritance (like `EntityList` → `Hand`/`Field`).

**Comment by andburn** [comment]
@Epix37 I'll get you the real frames :wink:

File: less/joust.less [hunk 1]
```diff
@@ -611,6 +611,70 @@
...
+			position: absolute;
+			height: 100%;
+			top: 5%;
+			left: 5%;
+		}
+
+		.text {
+			position: relative;
+			text-align: center;
+			top: -40%;
```
Comment (Epix37) [inline less/joust.less]: There has to be a better way of doing this, right?

File: less/joust.less [hunk 2]
```diff
@@ -611,6 +611,70 @@
...
+		position: relative;
+
+		.rank-portrait {
+			position: absolute;
+			height: 80%;
+			width: 98%;
+			top: 7%;
+			left: 6%;
+			clip-path: url(assets/clip-paths.svg#rank-clip);
+			-webkit-clip-path: polygon(50% 0%, 100% 25%, 100% 75%, 50% 100%, 0% 75%, 0% 25%);
```
Comment (beheh) [inline less/joust.less]: This has to be ported to SVG, for Firefox to work.

File: ts/protocol/HSReplayDecoder.ts [hunk 3]
```diff
@@ -177,7 +179,10 @@ class HSReplayDecoder extends Stream.Transform implements CardOracle {
 					id,
 					node.attributes.tags,
 					+node.attributes.playerID,
-					name
+					name,
+					undefined,
```
Comment (beheh) [inline ts/protocol/HSReplayDecoder.ts]: We can probably remove cardId from the player constructor (and just super `undefined`). I can do that later though.

File: ts/components/game/Rank.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+                    <RankArt legendRank={this.props.legendRank} assetDirectory={this.props.assetDirectory}/>
+                    <div className="text">
+                        {this.props.legendRank}
+                    </div>
+                </div>
+            );
+        }
+        else {
+            return (
+                <div className="rank no-entity"></div>
```
Comment (beheh) [inline ts/components/game/Rank.tsx]: Just return `null` here.

**Comment by beheh** [comment]
Ranks should be transferred in a reconnected replay, currently the rank vanishes, but the player name is kept (via playerMap in HSReplayDecoder). You could change player map from `number` to an object based on an interface (keep it in HSReplayDecoder):

```
interface PlayerDetails {
  id:number;
  rank?:number;
  legendRank?:number;
}
```

I don't mind if you want to do it or leave it open and I'll have a look at it later, we'll add an issue for it on merge then.

**Comment by Epix37** [comment]
As per suggestion on discord:
rankframe

**Comment by beheh** [comment]
Looking good! You can merge :+1:

**Pull request merged**